File: src/lib/memory/chunker.ts
import type { SessionTurn } from "../../types.js";
import type { Chunk } from "./types.js";
import { extractDateFromTurn } from "./normalizer.js";

const MAX_CHUNK_SIZE = 2000; // 每个 chunk 最大字符数
const CHUNK_OVERLAP = 200;  // 重叠字符数

/**
 * 把对话按语义分块
 */
export function chunkConversation(
  turns: SessionTurn[],
  sessionId: string
): Chunk[] {
  const chunks: Chunk[] = [];
  let currentChunk: string[] = [];
  let currentTurnIds: string[] = [];
  let currentSize = 0;
  let chunkIndex = 0;

  for (const turn of turns) {
    const turnText = `${turn.role}: ${turn.text}\n`;
    const turnSize = turnText.length;

    // 如果当前 chunk 加上这个 turn 超过限制，就保存当前 chunk
    if (currentSize + turnSize > MAX_CHUNK_SIZE && currentSize > 0) {
      chunks.push({
        id: `${sessionId}-chunk-${chunkIndex++}`,
        sourceSessionId: sessionId,
        sourceTurnIds: [...currentTurnIds],
        content: currentChunk.join(""),
        topic: "", // TODO: 后续用 LLM 生成主题
        metadata: {
          date: extractDateFromTurns(currentTurnIds, turns) || "",
          hasImages: currentTurnIds.some(id => turnHasImages(id, turns)),
          emotionalIntensity: "medium" // TODO: 后续实现情感强度识别
        }
      });

      // 保留重叠部分（最后 10% 的内容）
      const overlapLines = Math.max(1, Math.floor(currentChunk.length * 0.1));
      const overlap = currentChunk.slice(-overlapLines);
      const overlapTurnIds = currentTurnIds.slice(-overlapLines);

      currentChunk = overlap;
      currentTurnIds = overlapTurnIds;
      currentSize = overlap.join("").length;
    }

    currentChunk.push(turnText);
    currentTurnIds.push(turn.id);
    currentSize += turnSize;
  }

  // 保存最后一个 chunk
  if (currentChunk.length > 0) {
    chunks.push({
      id: `${sessionId}-chunk-${chunkIndex}`,
      sourceSessionId: sessionId,
      sourceTurnIds: currentTurnIds,
      content: currentChunk.join(""),
      topic: "",
      metadata: {
        date: extractDateFromTurns(currentTurnIds, turns) || "",
        hasImages: currentTurnIds.some(id => turnHasImages(id, turns)),
        emotionalIntensity: "medium"
      }
    });
  }

  return chunks;
}

function extractDateFromTurns(turnIds: string[], allTurns: SessionTurn[]): string | null {
  for (const id of turnIds) {
    const turn = allTurns.find(t => t.id === id);
    if (turn) {
      const date = extractDateFromTurn(turn);
      if (date) return date;
    }
  }
  return null;
}

function turnHasImages(turnId: string, allTurns: SessionTurn[]): boolean {
  return (allTurns.find(t => t.id === turnId)?.images.length ?? 0) > 0;
}

File: memory-system/src/chunker.py
from __future__ import annotations

from src.types import SessionTurn, Chunk
from src.normalizer import extract_date_from_turn

MAX_CHUNK_CHARS = 2000


def chunk_conversation(turns: list[SessionTurn], session_id: str) -> list[Chunk]:
    """把对话按大小分块，保留尾部重叠"""
    chunks: list[Chunk] = []
    buf_text: list[str] = []
    buf_ids: list[str] = []
    buf_size = 0
    idx = 0

    for turn in turns:
        turn_text = f"{turn.role}: {turn.text}\n"
        turn_size = len(turn_text)

        if buf_size + turn_size > MAX_CHUNK_CHARS and buf_size > 0:
            _flush_chunk(chunks, buf_text, buf_ids, session_id, idx, turns)
            idx += 1

            # 保留 10% 重叠
            overlap_n = max(1, len(buf_text) // 10)
            buf_text = buf_text[-overlap_n:]
            buf_ids = buf_ids[-overlap_n:]
            buf_size = sum(len(t) for t in buf_text)

        buf_text.append(turn_text)
        buf_ids.append(turn.id)
        buf_size += turn_size

    if buf_text:
        _flush_chunk(chunks, buf_text, buf_ids, session_id, idx, turns)

    return chunks


def _flush_chunk(
    chunks: list[Chunk],
    buf_text: list[str],
    buf_ids: list[str],
    session_id: str,
    idx: int,
    all_turns: list[SessionTurn],
) -> None:
    date = ""
    for tid in buf_ids:
        t = next((t for t in all_turns if t.id == tid), None)
        if t:
            d = extract_date_from_turn(t)
            if d:
                date = d
                break

    has_images = any(
        (t := next((t for t in all_turns if t.id == tid), None)) and t.images
        for tid in buf_ids
    )

    chunks.append(Chunk(
        id=f"{session_id}-chunk-{idx}",
        source_session_id=session_id,
        source_turn_ids=list(buf_ids),
        content="".join(buf_text),
        date=date,
        has_images=has_images,
    ))